python/ray/_private/gcs_utils.py

```diff
@@ -182,13 +196,227 @@ def internal_kv_keys(self, prefix: bytes) -> List[bytes]:
+                        self._errors.append((msg.key_id,
+                                             msg.error_info_message))
+
+            if len(self._errors) == 0:
+                return None, None
+            return self._errors.popleft()
+
+    def close(self) -> None:
+        """Closes the subscriber and its active subscriptions."""
+        # Mark close to terminate inflight polling and prevent future requests.
+        self._close.set()
+        with self._lock:
+            if not self._stub:
+                # Subscriber already closed.
+                return
+            req = gcs_service_pb2.GcsSubscriberCommandBatchRequest(
+                subscriber_id=self._subscriber_id, commands=[])
+            if self._subscribed_error:
+                cmd = pubsub_pb2.Command(
+                    channel_type=pubsub_pb2.RAY_ERROR_INFO_CHANNEL,
+                    unsubscribe_message={})
+                req.commands.append(cmd)
+            try:
+                self._stub.GcsSubscriberCommandBatch(req, timeout=30)
+            except Exception:
+                pass
+            self._stub = None
+
+
+class GcsAioPublisher:
+    """Publisher to GCS. Uses async io."""
+
+    def __init__(self, address: str = None, channel: grpc.aio.Channel = None):
+        if address:
+            assert channel is None, \
+                "address and channel cannot both be specified"
+            channel = create_gcs_aio_channel(address)
+        else:
+            assert channel is not None, \
+                "One of address and channel must be specified"
+        self._stub = gcs_service_pb2_grpc.InternalPubSubGcsServiceStub(channel)
+
+    async def publish_error(self, key_id: bytes,
+                            error_info: ErrorTableData) -> None:
+        """Publishes error info to GCS."""
+        msg = pubsub_pb2.PubMessage(
+            channel_type=pubsub_pb2.RAY_ERROR_INFO_CHANNEL,
+            key_id=key_id,
+            error_info_message=error_info)
+        req = gcs_service_pb2.GcsPublishRequest(pub_messages=[msg])
+        await self._stub.GcsPublish(req)
+
+
+class GcsAioSubscriber:
+    """Subscriber to GCS. Uses async io."""
+
+    def __init__(self, address: str = None, channel: grpc.aio.Channel = None):
+        if address:
+            assert channel is None, \
+                "address and channel cannot both be specified"
+            channel = create_gcs_aio_channel(address)
+        else:
+            assert channel is not None, \
+                "One of address and channel must be specified"
+        self._stub = gcs_service_pb2_grpc.InternalPubSubGcsServiceStub(channel)
+        self._subscriber_id = bytes(
+            bytearray(random.getrandbits(8) for _ in range(28)))
+        # Whether error info has been subscribed.
+        self._subscribed_error = False
+        # Buffer for holding error info.
+        self._errors = deque()
+
+    async def subscribe_error(self) -> None:
```

Comment:
Should this be a private method?

Reply:
This is part of the subscriber API. Updated comment. Calling `subscribe()` before polling is part of the flow for Redis pubsub and Ray pubsub C++ API. I think it is still needed for Ray pubsub python API, because it allows subscribing to a specific ID (even though it is not supported in pubsub python API yet). Also it ensures messages are accumulated before subscriber starts polling.